core/src/main/scala/kafka/zk/KafkaZkClient.scala

```
@@ -1599,4 +1690,28 @@ object KafkaZkClient {
       time, metricGroup, metricType)
     new KafkaZkClient(zooKeeperClient, isSecure, time)
   }
+
+
+  private def controllerZkVersionCheck(version: Int): Option[ZkVersionCheck] = {
+    if (version < 0)
+      None
+    else
+      Some(ZkVersionCheck(ControllerEpochZNode.path, version))
+  }
+
+  private def maybeThrowControllerMoveException(response: AsyncResponse): Unit = {
+    response.zkVersionCheckResult match {
+      case Some(zkVersionCheckResult) =>
+        val zkVersionCheck = zkVersionCheckResult.zkVersionCheck
+        if (zkVersionCheck.checkPath.equals(ControllerEpochZNode.path))
+          zkVersionCheckResult.opResult match {
+            case errorResult: ErrorResult =>
+              if (errorResult.getErr == Code.BADVERSION.intValue())
+              // Throw ControllerMovedException when the zkVersionCheck is performed on the controller epoch znode and the check fails
+                throw new ControllerMovedException(s"Controller epoch zkVersion check fails. Expected zkVersion = ${zkVersionCheck.expectedZkVersion}")
```

For errors other than BADVERSION, we should just propagate the original error as an exception.

---

Done.